src/contracts/ODSaviour.sol

```diff
@@ -1,12 +1,168 @@
+/**
+ * @notice Steps to save a safe using ODSaviour:
+ *
+ * 1. Protocol DAO => connect [this] saviour `LiquidationEngine.connectSAFESaviour`
+ * 2. Treasury DAO => enable specific vaults `ODSaviour.setVaultStatus`
+ * 3. Treasury DAO => approve `ERC20.approveTransferFrom` to the saviour
+ * 4. Vault owner => protect thier safe with elected saviour `LiquidationEngine.protectSAFE` (only works if ARB DAO enable vaultId)
+ * 5. Safe in liquidation => auto call `LiquidationEngine.attemptSave` gets saviour from chosenSAFESaviour mapping
+ * 6. Saviour => increases collateral `ODSaviour.saveSAFE`
+ */
+contract ODSaviour is AccessControl, IODSaviour {
+  using Math for uint256;
+  using Assertions for address;
 
-  function enableVault(uint256 _tokenId) external {}
+  // solhint-disable-next-line modifier-name-mixedcase
+  bytes32 public constant SAVIOUR_TREASURY = keccak256(abi.encode('SAVIOUR_TREASURY'));
+  bytes32 public constant PROTOCOL = keccak256(abi.encode('PROTOCOL'));
+
+  uint256 public liquidatorReward;
+
+  address public saviourTreasury;
+  address public protocolGovernor;
+  address public liquidationEngine;
+
+  IVault721 public vault721;
+  IOracleRelayer public oracleRelayer;
+  IODSafeManager public safeManager;
+  ISAFEEngine public safeEngine;
+  ICollateralJoinFactory public collateralJoinFactory;
+
+  mapping(uint256 _vaultId => bool _enabled) private _enabledVaults;
+  mapping(bytes32 _cType => IERC20 _tokenAddress) private _saviourTokenAddresses;
+
+  /**
+   * @param _init The SaviourInit struct;
+   */
+  constructor(SaviourInit memory _init) {
+    saviourTreasury = _init.saviourTreasury.assertNonNull();
+    protocolGovernor = _init.protocolGovernor.assertNonNull();
+    vault721 = IVault721(_init.vault721.assertNonNull());
+    oracleRelayer = IOracleRelayer(_init.oracleRelayer.assertNonNull());
+    safeManager = IODSafeManager(address(vault721.safeManager()));
+    liquidationEngine = ODSafeManager(address(safeManager)).liquidationEngine(); // todo update @opendollar package to include `liquidationEngine` - PR #693
+    collateralJoinFactory = ICollateralJoinFactory(_init.collateralJoinFactory.assertNonNull());
+    safeEngine = ISAFEEngine(address(safeManager.safeEngine()));
+    liquidatorReward = _init.liquidatorReward;
+
+    if (_init.saviourTokens.length != _init.cTypes.length) revert LengthMismatch();
+
+    // solhint-disable-next-line  defi-wonderland/non-state-vars-leading-underscore
+    for (uint256 i; i < _init.cTypes.length; i++) {
+      _saviourTokenAddresses[_init.cTypes[i]] = IERC20(_init.saviourTokens[i].assertNonNull());
+    }
+    _setupRole(SAVIOUR_TREASURY, saviourTreasury);
+    _setupRole(PROTOCOL, protocolGovernor);
+    _setupRole(PROTOCOL, liquidationEngine);
+  }
+
+  function isEnabled(uint256 _vaultId) external view returns (bool _enabled) {
+    _enabled = _enabledVaults[_vaultId];
+  }
+
+  function addCType(bytes32 _cType, address _tokenAddress) external onlyRole(SAVIOUR_TREASURY) {
```

Comment:
Update to use `_cParams` and `ModifiablePerCollateral` pattern as in the LiquidationEngine

```
  /// @inheritdoc ModifiablePerCollateral
  function _initializeCollateralType(bytes32 _cType, bytes memory _collateralParams) internal override whenEnabled {
    (LiquidationEngineCollateralParams memory _liqEngineCParams) =
      abi.decode(_collateralParams, (LiquidationEngineCollateralParams));
    _setCollateralAuctionHouse(_cType, _liqEngineCParams.collateralAuctionHouse);
    _cParams[_cType] = _liqEngineCParams;
  }

  /// @inheritdoc Modifiable
  function _modifyParameters(bytes32 _param, bytes memory _data) internal override {
    if (_param == 'onAuctionSystemCoinLimit') _params.onAuctionSystemCoinLimit = _data.toUint256();
    else if (_param == 'accountingEngine') accountingEngine = IAccountingEngine(_data.toAddress());
    else if (_param == 'saviourGasLimit') _params.saviourGasLimit = _data.toUint256();
    else revert UnrecognizedParam();
  }
  ```

Reply:
this is done on saviour branch